### ISSUE TITLE
f33: Possible race condition in the selinux fixup helper for stub-resolv.conf

### ISSUE THREAD

**Issue by dustymabe** [coreos/fedora-coreos-tracker#789]
**Describe the bug**

There is a possible race condition that has been exposed recently in CI. The selinux fixup helper for /run/systemd/resolve/stub-resolv.conf (introduced in https://github.com/coreos/fedora-coreos-config/commit/4356928, part of https://github.com/coreos/fedora-coreos-config/pull/780) is not working in some cases.

NOTE: Thus far this has only been seen when running kola tests on AWS.

The CI failures we are seeing look like:

```
=== RUN   ext.config.misc-ign-ro                                                                                                                                                                      
systemctl status kola-runext.service:                                                                                                                                                                 
● kola-runext.service                                                                                                                                                                                 
     Loaded: loaded (/etc/systemd/system/kola-runext.service; static)                                                                                                                                 
     Active: failed (Result: exit-code) since Thu 2021-04-08 18:34:41 UTC; 1s ago                                                                                                                     
    Process: 1911 ExecStart=/usr/local/bin/kola-runext-test.sh (code=exited, status=1/FAILURE)                                                                                                        
   Main PID: 1911 (code=exited, status=1/FAILURE)                                                                                                                                                     
                                                                                                                                                                                                      
Apr 08 18:34:40 ip-172-31-67-148 systemd[1]: Started kola-runext.service.                                                                                                                             
Apr 08 18:34:40 ip-172-31-67-148 kola-runext-test.sh[1911]: + grep -q zram0 /proc/swaps                                                                                                               
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: + ok 'swap on zram was set up correctly'                                                                                                  
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: + echo ok 'swap on zram was set up correctly'                                                                                             
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: ok swap on zram was set up correctly                                                                                                      
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1914]: + grep ActiveState=active                                                                                                                 
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1913]: + systemctl show coreos-update-ca-trust.service -p ActiveState                                                                            
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1914]: ActiveState=active                                                                                                                        
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: + grep '^# coreos.com$' /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt                                                           
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1915]: # coreos.com                                                                                                                              
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: + ok coreos-update-ca-trust.service                                                                                                       
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: + echo ok coreos-update-ca-trust.service                                                                                                  
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: ok coreos-update-ca-trust.service                                                                                                         
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1916]: ++ stat --format %C /run/systemd/resolve/stub-resolv.conf                                                                                 
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: + context=system_u:object_r:systemd_resolved_var_run_t:s0                                                                                 
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: + '[' system_u:object_r:systemd_resolved_var_run_t:s0 '!=' system_u:object_r:net_conf_t:s0 ']'                                            
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: + fatal 'SELinux context on stub-resolv.conf is wrong'                                                                                    
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: + echo 'SELinux context on stub-resolv.conf is wrong'                                                                                     
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: SELinux context on stub-resolv.conf is wrong                                                                                              
Apr 08 18:34:41 ip-172-31-67-148 kola-runext-test.sh[1911]: + exit 1
```

Indeed if you manage to run the test and get into the node the symlink has the wrong label:

```
[core@ip-172-31-67-148 ~]$ stat --format %C /run/systemd/resolve/stub-resolv.conf
system_u:object_r:systemd_resolved_var_run_t:s0
```

Though it's pretty clear that the path unit ran (see attached [journal.txt](https://github.com/coreos/fedora-coreos-tracker/files/6280881/journal.txt) log for more):

```
[core@ip-172-31-67-148 ~]$ journalctl -u coreos-reset-stub-resolv-selinux-context.service 
-- Logs begin at Thu 2021-04-08 18:34:05 UTC, end at Thu 2021-04-08 18:49:54 UTC. --
Apr 08 18:34:19 ip-172-31-67-148 systemd[1]: Starting Fixup SELinux context on /run/systemd/resolve/stub-resolv.conf...
Apr 08 18:34:19 ip-172-31-67-148 restorecon[950]: Relabeled /run/systemd/resolve/stub-resolv.conf from system_u:object_r:systemd_resolved_var_run_t:s0 to system_u:object_r:net_conf_t:s0
Apr 08 18:34:19 ip-172-31-67-148 systemd[1]: coreos-reset-stub-resolv-selinux-context.service: Succeeded.
Apr 08 18:34:19 ip-172-31-67-148 systemd[1]: Finished Fixup SELinux context on /run/systemd/resolve/stub-resolv.conf.
Apr 08 18:34:19 ip-172-31-67-148 systemd[1]: Starting Fixup SELinux context on /run/systemd/resolve/stub-resolv.conf...
Apr 08 18:34:19 ip-172-31-67-148 restorecon[1014]: Relabeled /run/systemd/resolve/stub-resolv.conf from system_u:object_r:systemd_resolved_var_run_t:s0 to system_u:object_r:net_conf_t:s0
Apr 08 18:34:19 ip-172-31-67-148 systemd[1]: coreos-reset-stub-resolv-selinux-context.service: Succeeded.
Apr 08 18:34:19 ip-172-31-67-148 systemd[1]: Finished Fixup SELinux context on /run/systemd/resolve/stub-resolv.conf.
```

I guess it's possible that the path units can not get triggered on every access to the file maybe if the triggered unit was already running. Not sure. 

**Reproduction steps**

If I run the following a few times it seems to be able to reproduce:

```
cosa kola run --ssh-on-test-failure -p aws --aws-credentials-file=aws.ini --aws-ami ami-025f6060a1ea16d97 --aws-region us-east-1 -b fcos -j 5 --no-test-exit-error ext.config.misc-ign-ro
```

**Comment by dustymabe** [comment]
I marked this with `f33` because we don't need the selinux fixup hack in f34.

**Comment by dustymabe** [comment]
More info.. Since we saw this happen on the kola-aws run for the stable release we decided to dig in a bit more:

- only observed on AWS
- only observed when running via `kola` (either `kola run` or `kola spawn`)

Considering the reset of the label is admittedly a hack, I'm not suprised there is a very subtle race here that seems to be only triggered when being brought up by kola on AWS. This hack will be gone automatically when we rebase the various streams to F34. Let's hope we don't see it anywhere else.
